ext/storage_sources/azure_store/test/test_azure_connection.cpp

```diff
@@ -30,5 +30,281 @@
+    {
+        azure_connection conn = azure_connection("myblobcontainer1", "delete_obj_");
+
+        auto blob_client = azure_client.GetBlockBlobClient("delete_obj_" + file_name);
+
+        blob_client.UploadFrom(file_name);
+
+        // Test that an object can be deleted.
+        CHECK(conn.delete_object(file_name) == 0);
+
+        // Test that removing an object that doesn't exist returns a ENOENT.
+        CHECK(conn.delete_object(non_exi_object_key) == ENOENT);
+    }
+
+    SECTION("(Checking put functionality in Azure)", "[azure-connection]")
+    {
+        azure_connection conn = azure_connection("myblobcontainer1", "put_obj_");
+
+        CHECK(conn.put_object(file_name, path) == 0);
+
+        auto blob_client = azure_client.GetBlockBlobClient("put_obj_" + file_name);
+        blob_client.Delete();
+
+        // Test that putting an object that doesn't exist locally returns -1.
+        CHECK(conn.put_object(non_exi_object_key, non_exi_file_name) == -1);
+    }
+
+    SECTION("(Checking read functionality in Azure)", "[azure-connection]")
+    {
+        azure_connection conn = azure_connection("myblobcontainer1", "read_obj_");
+
+        auto blob_client = azure_client.GetBlockBlobClient("read_obj_" + file_name);
+
+        blob_client.UploadFrom(file_name);
+        void *buffer = calloc(1024, sizeof(char));
+
+        // Test reading whole file.
+        CHECK(conn.read_object(file_name, 0, payload.length(), buffer) == 0);
+        CHECK(static_cast<char *>(buffer) == payload);
+        memset(buffer, 0, 1024);
+
+        // Test overflow on positive offset but past EOF.
+        CHECK(conn.read_object(file_name, 1, 1000, buffer) == -1);
+        memset(buffer, 0, 1024);
+
+        // Test overflow on negative offset but past EOF.
+        CHECK(conn.read_object(file_name, -1, 1000, buffer) == -1);
+        memset(buffer, 0, 1024);
+
+        // Test overflow with negative offset.
+        CHECK(conn.read_object(file_name, -1, 12, buffer) == -1);
+        free(buffer);
+
+        blob_client.Delete();
+
+        // Tests that reading a non existant object returns a ENOENT.
+        CHECK(conn.read_object(non_exi_object_key, 0, 1, buffer) == ENOENT);
+    }
+
+    SECTION("(Test complex case).", "[azure-connection]")
+    {
+        azure_connection conn = azure_connection("myblobcontainer1", "complex_case_");
+        azure_connection conn_bad = azure_connection("myblobcontainer1", "bad_prefix_");
+
+        // Read an object in the container from middle to end
+        // Prefix for objects in this test.
+        const std::string prefix = "complex_case_";
+        const bool list_single = true;
+        void *buffer = calloc(1024, sizeof(char));
+        std::vector<std::string> objects;
+
+        // Total objects to insert in the test.
+        const int32_t num_objects = 11;
+
+        // Check that the object does not exist in the container.
+        CHECK(conn.object_exists(file_name_2, exists) == 0);
```

Comment:
I'm having second thoughts of having a complex case if we are already testing each function individually. What are your thoughts? Do we have added benefit with this unit test?

Reply:
I agree - When I first wrote this test testing complex test for each function individually and in isolation it seems like there was a lot of repetition. This is the same with the current complex case so I'm in favour of removing the complex test case provided we have considered all cases in the individual ones.

Reply:
removed